app/proxy/proxy.go

```diff
@@ -0,0 +1,330 @@
+			return nil, errors.Wrap(err, "failed to create proxy")
+		}
+		pc.conf.listener = l
+		err = pc.configureBackend(menderUrl, menderJwtToken)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create proxy")
+		}
+		pc.Start()
+	}
+
+	return pc, nil
+}
+
+func newNetListener() (net.Listener, error) {
+	l, err := net.Listen("tcp", ":0")
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create listener")
+	}
+	return l, nil
+}
+
+func (pc *ProxyController) configureBackend(menderUrl, menderJwtToken string) error {
+	u, err := url.Parse(menderUrl)
+	if err != nil {
+		return errors.Wrap(err, "failed to configure proxy")
+	}
+
+	pc.conf.backend = u
+	pc.conf.jwtToken = menderJwtToken
+	return nil
+}
+
+// Reconfigure reconfigures the local proxy server
+func (pc *ProxyController) Reconfigure(menderUrl, menderJwtToken string) error {
+	if pc.isRunning {
+		return errors.New("failed to reconfigure proxy: cannot reconfigure while running")
+	}
+
+	l, err := newNetListener()
+	if err != nil {
+		return errors.Wrap(err, "failed to reconfigure proxy")
+	}
+	pc.conf.listener = l
+
+	err = pc.configureBackend(menderUrl, menderJwtToken)
+	if err != nil {
+		return errors.Wrap(err, "failed to reconfigure proxy")
+	}
+
+	return nil
+}
+
+func (pc *ProxyController) getPort() int {
+	return pc.conf.listener.Addr().(*net.TCPAddr).Port
+}
+
+// GetServerUrl returns the URL of the proxy
+func (pc *ProxyController) GetServerUrl() string {
+	if pc.isRunning {
+		return fmt.Sprintf("http://%s:%d", ProxyHost, pc.getPort())
+	}
+	return ""
+}
+
+func copyResponse(rw http.ResponseWriter, resp *http.Response) error {
+	copyHeader(rw.Header(), resp.Header)
+	rw.WriteHeader(resp.StatusCode)
+	defer resp.Body.Close()
+
+	_, err := io.Copy(rw, resp.Body)
+	return err
+}
+
+func copyHeader(dst, src http.Header) {
+	for k, vv := range src {
+		for _, v := range vv {
+			dst.Add(k, v)
+		}
+	}
+}
+
+// Start starts the local proxy server
+func (pc *ProxyController) Start() {
+	if pc.isRunning {
+		return
+	}
+
+	// TODO: how to ensure it has a listener before Start?
+	if pc.conf.listener != nil {
+		pc.isRunning = true
+
+		initDone := make(chan struct{}, 1)
+		go pc.run(initDone)
+
+		// Wait for the server to start
+		<-initDone
+
+		runtime.SetFinalizer(pc, func(pc *ProxyController) {
+			pc.Stop()
+		})
+	}
+}
+
+// Stop stops the local proxy server
+func (pc *ProxyController) Stop() {
+	if !pc.isRunning {
+		return
+	}
+
+	if pc.wsRunning() {
+		pc.CloseWsConnections()
+	}
+
+	pc.quitReq <- struct{}{}
+	// Wait for server to shutdown
+	<-pc.quitResp
+	pc.isRunning = false
+
+	runtime.SetFinalizer(pc, nil)
+}
+
+func (pc *ProxyController) run(initDone chan struct{}) {
+	mux := http.NewServeMux()
+	mux.HandleFunc(ApiUrlDevicesPrefix, pc.checkAuthorizationHook(pc.doHttpRequest))
+	mux.HandleFunc(ApiUrlDevicesAuthentication, pc.apiDevicesAuthenticationHandler)
+	mux.HandleFunc(ApiUrlDevicesConnect, pc.checkAuthorizationHook(pc.apiDevicesConnectHandler))
```

Comment:
websockets protocol does not use `CONNECT` method. It works with a `GET` that includes headers `Connection: Upgrade` and `Upgrade: websocket`, to which the server responds with code `101 Switching Protocols` and both parties agreed then to keep the TCP connection alive.

Mozilla [docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Protocol_upgrade_mechanism) are simple but pretty good.

I haven't read the RFC in detail, but I think there is no way to distinguish beforehand because even this `ws://` or `wss://` is a convenience thing, you can actually use `http://` and `https://` and include such headers.

In short, either we have a hook that checks for this headers and then passes to one handler or the other, or we leave it as is and add individual handlers for individual endpoints as we go...

Reply:
You're right, I confused websockets and CONNECT. Never mind then.